platform/linux-generic/odp_ipsec.c

```diff
@@ -1928,6 +1965,9 @@ int odp_ipsec_result(odp_ipsec_packet_result_t *result, odp_packet_t packet)
 	/* FIXME: maybe postprocess here, setting alg error in case of crypto
 	 * error instead of processing packet fully in ipsec_in/out_single */
 
+	if (ipsec_config->stats_en)
+		ipsec_sa_err_stats_update(res);
```

Comment:
Doing the update here means that statistics will not get updated if the application does not call odp_ipsec_result(). I think the update should be done in ipsec_in_single() and in ipsec_out_single().

Reply:
I did ask the same question myself while making this change. For linux-generic it should be fine either way, but with implementations having hardware accelerators, it may not be possible to do it that way (as in, state of the operation may not be established by the time the API returns). So ipsec result check would be kind of mandatory in such cases. 

But for linux-generic, I don't mind either way. Please let me know your preference and I can rework.

Reply:
The problem is that the API does not mandate that the application calls odp_ipsec_result() right after IPsec processing or ever. In practise it might be a rare case that the application never calls the function but I can imagine that some application may e.g. queue the packet and call the result function much later in another thread.

So the question is, what is the state of the packet or event returned by your IPsec implementation before odp_ipsec_result() gets called? Does skipping the odp_ipsec_result() call screw up merely the statistics or something else too?

Why cannot IPsec processing be fully done already when the application gets the result packet? Too costly to check for IPsec processing completion need for every packet in some generic scheduling or dequeueing functions so you would like to postpone that to odp_ipsec_result()?

If you really require that odp_ipsec_result() is always called or bad things (or just statistics miscounts) will happen, then I think the API needs to be updated.

Reply:
> 
> The problem is that the API does not mandate that the application calls odp_ipsec_result() right after IPsec processing or ever. In practise it might be a rare case that the application never calls the function but I can imagine that some application may e.g. queue the packet and call the result function much later in another thread.
> 
> So the question is, what is the state of the packet or event returned by your IPsec implementation before odp_ipsec_result() gets called? Does skipping the odp_ipsec_result() call screw up merely the statistics or something else too?

Isn't IPsec result check mandatory? Like, in case of ICV errors, it is an auditable event.

https://tools.ietf.org/html/rfc4303#page-30

> 
> Why cannot IPsec processing be fully done already when the application gets the result packet? Too costly to check for IPsec processing completion need for every packet in some generic scheduling or dequeueing functions so you would like to postpone that to odp_ipsec_result()?

Yes. It would make the scheduling/dequeuing functions to grow and become rather inefficient for non-IPsec use cases.

> 
> If you really require that odp_ipsec_result() is always called or bad things (or just statistics miscounts) will happen, then I think the API needs to be updated.

From the RFC it looks like, result check is mandatory. So do you recommend documenting in the API? I can submit the patch if you suggest so.

Reply:
> Isn't IPsec result check mandatory? Like, in case of ICV errors, it is an auditable event.

IPsec RFC compliancy matters do not tell how the application is allowed to use the API and one cannot deduce that based on certain RFC clause an ODP application must always all odp_ipsec_result(). Maybe the application does not even want to be RFC compliant? Maybe the application uses odp_packet_has_error() to check when it needs to call odp_ipsec_result() (yes, there is ongoing discussion about how odp_packet_has_error() should work).

> So do you recommend documenting in the API? I can submit the patch if you suggest so.

The API needs to fully document how it must or must not be used, so please suggest a change if you think it is needed. But try to define very carefully when exactly the application must call odp_ipsec_result() and what happens if it does not (undefined behaviour / inaccurate statistics / some functions may not be called for the packet). I think one use case that probably needs to work is reception of IPsec events and then freeing them without any processing at all.

Reply:
> 
> 
> > Isn't IPsec result check mandatory? Like, in case of ICV errors, it is an auditable event.
> 
> IPsec RFC compliancy matters do not tell how the application is allowed to use the API and one cannot deduce that based on certain RFC clause an ODP application must always all odp_ipsec_result(). Maybe the application does not even want to be RFC compliant? Maybe the application uses odp_packet_has_error() to check when it needs to call odp_ipsec_result() (yes, there is ongoing discussion about how odp_packet_has_error() should work).

Understood. Should I update the ipsec_stats documentation to reflect that application is expected to call odp_ipsec_result() for stats to be updated properly? Putting it in schedule/dequeue APIs will mean non-IPsec packets will have an unnecessary check and I would like to avoid that.

> 
> > So do you recommend documenting in the API? I can submit the patch if you suggest so.
> 
> The API needs to fully document how it must or must not be used, so please suggest a change if you think it is needed. But try to define very carefully when exactly the application must call odp_ipsec_result() and what happens if it does not (undefined behaviour / inaccurate statistics / some functions may not be called for the packet). I think one use case that probably needs to work is reception of IPsec events and then freeing them without any processing at all.

Agreed. IMO availability of APIs like odp_ipsec_result() allows schedule/dequeue APIs to be light while allowing various module specific processing to be done separately. Let me recheck the documentation. I can submit API documentation changes if required. For this case, I can update the ipsec_stats documentation to reflect the dependency on odp_ipsec_result().

Reply:
@JannePeltonen Since the API for post-processing may need some more iterations, shall we merge this now and I can take up moving stats collection to the API that we decide for post-processing, once we agree on that PR.

Reply:
For us to merge this before changing the API to allow post processing we need to move the ipsec_sa_err_stats_update() call from odp_ipsec_result() to somewhere like ipsec_in_single() and ipsec_out_single(). I think we should do that change and try to get this merged soon.